src/EventStore.Client.Streams/EventStoreClient.Read.cs

```diff
@@ -110,130 +208,144 @@ public partial class EventStoreClient {
+			/// </summary>
+			public StreamPosition? LastStreamPosition { get; private set; }
 
-					_moved = await call.MoveNextAsync(cancellationToken).ConfigureAwait(false);
-					return call.Current?.ContentCase switch {
-						ReadResp.ContentOneofCase.StreamNotFound => Client.ReadState.StreamNotFound,
-						_ => Client.ReadState.Ok
-					};
+			/// <summary>
+			/// An <see cref="IAsyncEnumerable{StreamMessage}"/>.
+			/// </summary>
+			public IAsyncEnumerable<StreamMessage> Messages {
+				get {
+					return GetMessages();
+
+					async IAsyncEnumerable<StreamMessage> GetMessages() {
+						await foreach (var message in _channel.Reader.ReadAllAsync().ConfigureAwait(false)) {
+							switch (message) {
+								case StreamMessage.FirstStreamPosition(var streamPosition):
+									FirstStreamPosition = streamPosition;
+									break;
+								case StreamMessage.LastStreamPosition(var lastStreamPosition):
+									LastStreamPosition = lastStreamPosition;
+									break;
+								default:
+									break;
+							}
+
+							yield return message;
+						}
+					}
 				}
-
-				Current = default;
 			}
 
 			/// <summary>
 			/// The <see cref="ReadState"/>.
 			/// </summary>
 			public Task<ReadState> ReadState { get; }
 
-			/// <inheritdoc />
-			public IAsyncEnumerator<ResolvedEvent> GetAsyncEnumerator(
-				CancellationToken cancellationToken = new CancellationToken()) {
-				_cancellationToken = cancellationToken;
-				return this;
-			}
-
-			/// <inheritdoc />
-			public async ValueTask DisposeAsync() =>
-				await (await _callSource.Task.ConfigureAwait(false)).DisposeAsync().ConfigureAwait(false);
-
-			/// <inheritdoc />
-			public async ValueTask<bool> MoveNextAsync() {
-				var state = await ReadState.ConfigureAwait(false);
-				if (state != Client.ReadState.Ok) {
-					throw ExceptionFromState(state, _streamName);
-				}
+			internal ReadStreamResult(Func<CancellationToken, Task<CallInvoker>> selectCallInvoker, ReadReq request,
+				EventStoreClientSettings settings, EventStoreClientOperationOptions operationOptions,
+				UserCredentials? userCredentials, CancellationToken cancellationToken) {
+				var callOptions = EventStoreCallOptions.Create(settings, operationOptions, userCredentials,
+					cancellationToken);
 
-				var call = await _callSource.Task.ConfigureAwait(false);
+				_channel = Channel.CreateBounded<StreamMessage>(new BoundedChannelOptions(1) {
+					SingleReader = true,
+					SingleWriter = true,
+					AllowSynchronousContinuations = true
+				});
 
-				if (_moved) {
-					_moved = false;
-					if (IsCurrentItemEvent()) {
-						return true;
+				StreamName = request.Options.Stream.StreamIdentifier;
+
+				var tcs = new TaskCompletionSource<ReadState>();
+#pragma warning disable CS0612
+				ReadState = tcs.Task;
+#pragma warning restore CS0612
+
+				PumpMessages();
+
+				async void PumpMessages() {
+					var callInvoker = await selectCallInvoker(cancellationToken).ConfigureAwait(false);
```

Comment:
are these intentionally outside of the try? a bit nervous about the possibility of throwing exceptions out of the async void (same in all stream version)